### PR TITLE
Guard against negative result from FileStore.getUsableSpace when picking data path for a new shard

### PR DESCRIPTION
**Pull request by mikemccand**
A user in https://discuss.elastic.co/t/elasticsearch-with-amazon-elastic-file-system/55867/5 hit a `NullPointerException` in `ShardPath.selectNewPathForShard`, I think because `FileStore.getUsableSpace` returned `Long.MIN_VALUE`.

So I fixed `ShardPath` to `Math.max` the returned result with 0, and added a test case showing the NPE.

### CODE REVIEW

**Comment by nik9000** [comment]
LGTM

**Comment by rmuir** [comment]
are we sure its correct to treat them as 0?

FileStore computes this method as: `return attrs.blockSize() * attrs.availableBlocks()`

both of those attributes are java `long`, but the values coming from statvfs() are `unsigned long` types on linux. So how can they be negative :)

**Comment by rmuir** [comment]
Also i notice there are talks on the internet about EFS with names such as "how to store 8 exabytes and look good doing it". Perhaps it is a simple case of overflow?

**Comment by mikemccand** [comment]
> are we sure its correct to treat them as 0?

Well, for this `ShardPath` method in particular, since we are just using it to pick which `path.data` to use for this shard, I think using `0` instead is fine.

But in general, e.g. in `ESFileStore`, I don't think we should do that.

I could also fix this w/o futzing with the invalid usable space, by initializing `bestPath` to `paths[0]` or skip the loop entirely if there's only one `path.data` ... that would avoid this NPE.

Really this is all best-effort: the fact is, there is a bug-in-something (your JVM, your filesystem), if it's returning negative values from `FileStore.getUsableSpace`.

**Comment by rmuir** [comment]
> Well, for this ShardPath method in particular, since we are just using it to pick which path.data to use for this shard, I think using 0 instead is fine.
> 
> But in general, e.g. in ESFileStore, I don't think we should do that.
> 
> I could also fix this w/o futzing with the invalid usable space, by initializing bestPath to paths[0] or skip the loop entirely if there's only one path.data ... that would avoid this NPE.
> 
> Really this is all best-effort: the fact is, there is a bug-in-something (your JVM, your filesystem), if it's returning negative values from FileStore.getUsableSpace.

I don't think it works here. It does not fix the problem and it will only be encountered again, in other parts of the code. Setting it to 0 doesn't seem right, besides that, if free space can overflow, then other things like total space probably overflow too, and all the crazy math here (which we should not be doing, unless we use BigInteger) is probably going insane too. Looks like a big pile of bugs in ES here, too, to me!

Instead, we should fix the bugs. The first thing to do is to get the output of `df` for this filesystem so we can discuss it with the jdk developers, and figure out how they plan to fix the multiplication in FileStore. For example, if they plan to expose new methods that return BigInteger, versus attributes returning block size and number of blocks that should be treated as unsigned longs, or whatever. Depending on what they are doing, any workaround (if we should even have one), should be done in a way that will be easy to remove when its no longer necessary.

Until then, I am fine with NFS NPE'ing, kind of a bonus feature.

**Comment by rmuir** [comment]
By the way, on every OS we support (linux, os X, and windows) the underlying OS values are unsigned longs. Negative values are not possible. So to me its no real mystery what happens, but the `df` output will just make the conversation easier so we can move forward to get it fixed.

**Comment by mikemccand** [comment]
OK I've asked the user for `df` output in their case: https://discuss.elastic.co/t/elasticsearch-with-amazon-elastic-file-system/55867/8?u=mikemccand

**Comment by mikemccand** [comment]
I removed the `feedback_needed` label: the user replied and indeed this is a bug in the JDK: https://bugs.openjdk.java.net/browse/JDK-8162520

In that issue they are leaning to having the `FileStore` APIs returning `long` just return `Long.MAX_VALUE` if they would otherwise overflow ... so I did that in this PR, when a negative value is returned.

I also upgraded the `ShardPath` logic to use `BigInteger` and made it a bit more defensive.

And I changed the test case to be even more unit-y, testing `ESFileStore` directly.

I think it's ready.

**Comment by rmuir** [comment]
+1, this is the most reasonable solution for now.

File: core/src/test/java/org/elasticsearch/env/ESFileStoreTests.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        @Override
+        public Object getAttribute(String attribute) throws IOException {
+            return null;
+        }
+    }
+
+    public void testNegativeSpace() throws Exception {
+        FileStore store = new ESFileStore(new MockFileStore());
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/env/ESFileStoreTests.java]: You can use mockito here if you like, I imagine.

**Comment by nik9000** [comment]
Makes sense to me as well. Sorry for jumping the gun on this one.

File: core/src/test/java/org/elasticsearch/env/ESFileStoreTests.java (hunk 1 above)
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/env/ESFileStoreTests.java]: > You can use mockito here if you like, I imagine.

Indeed, so much simpler!  Thank you for the pointer.

**Comment by nik9000** [comment]
LGTM

**Comment by mikemccand** [comment]
I agree it makes sense to backport this for a future 2.4.x release @olegmyrk; I'll open a PR.

**Pull request merged**